File: 2021/Python/day01.py


def solve1() -> int:
    with open('../.input/day01', 'r') as f:
        numbers: list[int] = [int(line) for line in f.readlines()]
    return sum((n > p) for p, n in zip(numbers[:-1], numbers[1:]))


def solve2() -> int:
    with open('../.input/day01', 'r') as f:
        numbers: list[int] = [int(line) for line in f.readlines()]
    return sum((sum(n) > sum(p)) for p, n in zip(
        zip(numbers[:-3], numbers[1:-2], numbers[2:-1]),
        zip(numbers[1:-2], numbers[2:-1], numbers[3:])
    ))
    

if __name__ == '__main__':
    print(solve1())  # 1766
    print(solve2())  # 1797

File: 2021/rust/src/solutions/day01.rs
use crate::utils;
use crate::utils::Source;


pub fn run() -> () {
    let data = parse_data(utils::read_lines(Source::Day(1)));

    println!("Day 1");
    println!("Part 1: {}", solve1(&data));
    println!("Part 2: {}", solve2(&data));
}

fn solve1(lines: &Vec<i32>) -> i32 {
    let mut increments = 0;
    let mut prev = i32::MAX;
    for &n in lines {
        if n > prev { increments += 1 };
        prev = n;
    };
    increments
}

fn solve2(lines: &Vec<i32>) -> i32 {
    let length = lines.len();

    let left = lines[..length - 1].windows(3);
    let right = lines[1..].windows(3);


    left.zip(right)
        .fold(0, |acc, (left, right)| {
            let sum_left: i32 = left.iter().sum();
            let sum_right: i32 = right.iter().sum();
            acc + if sum_left < sum_right { 1 } else { 0 }
        })
}

fn parse_data(lines: Vec<String>) -> Vec<i32> {
    lines.iter()
        .map(|x| x.parse())
        .collect::<Result<Vec<i32>, _>>()
        .expect("Failed to parse int")
}
